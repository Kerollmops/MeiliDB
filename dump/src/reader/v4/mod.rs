use std::{
    fs::{self, File},
    io::{BufRead, BufReader},
    path::Path,
};

use serde::{Deserialize, Serialize};
use tempfile::TempDir;
use time::OffsetDateTime;
use uuid::Uuid;

mod keys;
mod meta;
mod settings;
mod tasks;

use crate::{IndexMetadata, Result, Version};

use self::{
    keys::Key,
    meta::{DumpMeta, IndexUuid},
    settings::{Checked, Settings},
    tasks::Task,
};

use super::{/* compat::v4_to_v5::CompatV4ToV5, */ DumpReader, IndexReader};

pub type Document = serde_json::Map<String, serde_json::Value>;
pub type UpdateFile = File;

#[derive(Serialize, Deserialize, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Metadata {
    db_version: String,
    index_db_size: usize,
    update_db_size: usize,
    #[serde(with = "time::serde::rfc3339")]
    dump_date: OffsetDateTime,
}

pub struct V4Reader {
    dump: TempDir,
    metadata: Metadata,
    tasks: BufReader<File>,
    keys: BufReader<File>,
    index_uuid: Vec<IndexUuid>,
}

impl V4Reader {
    pub fn open(dump: TempDir) -> Result<Self> {
        let meta_file = fs::read(dump.path().join("metadata.json"))?;
        let metadata = serde_json::from_reader(&*meta_file)?;
        let index_uuid = File::open(dump.path().join("index_uuids/data.jsonl"))?;
        let index_uuid = BufReader::new(index_uuid);
        let index_uuid = index_uuid
            .lines()
            .map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) })
            .collect::<Result<Vec<_>>>()?;

        Ok(V4Reader {
            metadata,
            tasks: BufReader::new(
                File::open(dump.path().join("updates").join("data.jsonl")).unwrap(),
            ),
            keys: BufReader::new(File::open(dump.path().join("keys"))?),
            index_uuid,
            dump,
        })
    }

    // pub fn to_v5(self) -> CompatV4ToV5 {
    //     CompatV4ToV5::new(self)
    // }

    pub fn version(&self) -> Version {
        Version::V4
    }

    pub fn date(&self) -> Option<OffsetDateTime> {
        Some(self.metadata.dump_date)
    }

    pub fn instance_uid(&self) -> Result<Option<Uuid>> {
        let uuid = fs::read_to_string(self.dump.path().join("instance-uid"))?;
        Ok(Some(Uuid::parse_str(&uuid)?))
    }

    pub fn indexes(&self) -> Result<impl Iterator<Item = Result<V4IndexReader>> + '_> {
        Ok(self.index_uuid.iter().map(|index| -> Result<_> {
            Ok(V4IndexReader::new(
                index.uid.clone(),
                &self
                    .dump
                    .path()
                    .join("indexes")
                    .join(index.index_meta.uuid.to_string()),
            )?)
        }))
    }

    pub fn tasks(&mut self) -> impl Iterator<Item = Result<(Task, Option<UpdateFile>)>> + '_ {
        (&mut self.tasks).lines().map(|line| -> Result<_> {
            let task: Task = serde_json::from_str(&line?)?;
            if !task.is_finished() {
                if let Some(uuid) = task.get_content_uuid() {
                    let update_file_path = self
                        .dump
                        .path()
                        .join("updates")
                        .join("updates_files")
                        .join(uuid.to_string());
                    Ok((task, Some(File::open(update_file_path).unwrap())))
                } else {
                    Ok((task, None))
                }
            } else {
                Ok((task, None))
            }
        })
    }

    pub fn keys(&mut self) -> impl Iterator<Item = Result<Key>> + '_ {
        (&mut self.keys)
            .lines()
            .map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) })
    }
}

pub struct V4IndexReader {
    metadata: IndexMetadata,
    settings: Settings<Checked>,

    documents: BufReader<File>,
}

impl V4IndexReader {
    pub fn new(name: String, path: &Path) -> Result<Self> {
        let meta = File::open(path.join("meta.json"))?;
        let meta: DumpMeta = serde_json::from_reader(meta)?;

        let metadata = IndexMetadata {
            uid: name,
            primary_key: meta.primary_key,
            // FIXME: Iterate over the whole task queue to find the creation and last update date.
            created_at: OffsetDateTime::now_utc(),
            updated_at: OffsetDateTime::now_utc(),
        };

        let ret = V4IndexReader {
            metadata,
            settings: meta.settings.check(),
            documents: BufReader::new(File::open(path.join("documents.jsonl"))?),
        };

        Ok(ret)
    }

    pub fn metadata(&self) -> &IndexMetadata {
        &self.metadata
    }

    pub fn documents(&mut self) -> Result<impl Iterator<Item = Result<Document>> + '_> {
        Ok((&mut self.documents)
            .lines()
            .map(|line| -> Result<_> { Ok(serde_json::from_str(&line?)?) }))
    }

    pub fn settings(&mut self) -> Result<Settings<Checked>> {
        Ok(self.settings.clone())
    }
}

#[cfg(test)]
pub(crate) mod test {
    use std::{fs::File, io::BufReader};

    use flate2::bufread::GzDecoder;
    use tempfile::TempDir;

    use super::*;

    #[test]
    fn read_dump_v4() {
        let dump = File::open("tests/assets/v4.dump").unwrap();
        let dir = TempDir::new().unwrap();
        let mut dump = BufReader::new(dump);
        let gz = GzDecoder::new(&mut dump);
        let mut archive = tar::Archive::new(gz);
        archive.unpack(dir.path()).unwrap();

        let mut dump = V4Reader::open(dir).unwrap();

        // top level infos
        insta::assert_display_snapshot!(dump.date().unwrap(), @"2022-10-06 12:53:49.131989609 +00:00:00");
        insta::assert_display_snapshot!(dump.instance_uid().unwrap().unwrap(), @"9e15e977-f2ae-4761-943f-1eaf75fd736d");

        // tasks
        let tasks = dump.tasks().collect::<Result<Vec<_>>>().unwrap();
        let (tasks, update_files): (Vec<_>, Vec<_>) = tasks.into_iter().unzip();
        insta::assert_json_snapshot!(tasks);
        assert_eq!(update_files.len(), 10);
        assert!(update_files[0].is_some()); // the enqueued document addition
        assert!(update_files[1..].iter().all(|u| u.is_none())); // everything already processed

        // keys
        let keys = dump.keys().collect::<Result<Vec<_>>>().unwrap();
        insta::assert_json_snapshot!(keys);

        // indexes
        let mut indexes = dump.indexes().unwrap().collect::<Result<Vec<_>>>().unwrap();
        // the index are not ordered in any way by default
        indexes.sort_by_key(|index| index.metadata().uid.to_string());

        let mut products = indexes.pop().unwrap();
        let mut movies = indexes.pop().unwrap();
        let mut spells = indexes.pop().unwrap();
        assert!(indexes.is_empty());

        // products
        insta::assert_json_snapshot!(products.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "products",
          "primaryKey": "sku",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(products.settings());
        let documents = products
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 10);
        insta::assert_json_snapshot!(documents);

        // movies
        insta::assert_json_snapshot!(movies.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "movies",
          "primaryKey": "id",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(movies.settings());
        let documents = movies
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 110);
        insta::assert_debug_snapshot!(documents);

        // spells
        insta::assert_json_snapshot!(spells.metadata(), { ".createdAt" => "[now]", ".updatedAt" => "[now]" }, @r###"
        {
          "uid": "dnd_spells",
          "primaryKey": "index",
          "createdAt": "[now]",
          "updatedAt": "[now]"
        }
        "###);

        insta::assert_debug_snapshot!(spells.settings());
        let documents = spells
            .documents()
            .unwrap()
            .collect::<Result<Vec<_>>>()
            .unwrap();
        assert_eq!(documents.len(), 10);
        insta::assert_json_snapshot!(documents);
    }
}
