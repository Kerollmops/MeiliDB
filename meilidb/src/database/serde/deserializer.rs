use std::error::Error;
use std::ops::Deref;
use std::fmt;

use rocksdb::rocksdb::{DB, Snapshot, SeekKey};
use rocksdb::rocksdb_options::ReadOptions;
use serde::forward_to_deserialize_any;
use serde::de::value::MapDeserializer;
use serde::de::{self, Visitor, IntoDeserializer};

use crate::database::document_key::{DocumentKey, DocumentKeyAttr};
use crate::database::schema::Schema;
use meilidb_core::DocumentId;

pub struct Deserializer<'a, D>
where D: Deref<Target=DB>
{
    snapshot: &'a Snapshot<D>,
    schema: &'a Schema,
    document_id: DocumentId,
}

impl<'a, D> Deserializer<'a, D>
where D: Deref<Target=DB>
{
    pub fn new(snapshot: &'a Snapshot<D>, schema: &'a Schema, doc: DocumentId) -> Self {
        Deserializer { snapshot, schema, document_id: doc }
    }
}

impl<'de, 'a, 'b, D> de::Deserializer<'de> for &'b mut Deserializer<'a, D>
where D: Deref<Target=DB>
{
    type Error = DeserializerError;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.deserialize_map(visitor)
    }

    forward_to_deserialize_any! {
        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string unit seq
        bytes byte_buf unit_struct tuple_struct
        identifier tuple ignored_any option newtype_struct enum struct
    }

    fn deserialize_map<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        let mut options = ReadOptions::new();
        let lower = DocumentKey::new(self.document_id);
        let upper = lower.with_attribute_max();
        options.set_iterate_lower_bound(lower.as_ref());
        options.set_iterate_upper_bound(upper.as_ref());

        let mut iter = self.snapshot.iter_opt(options);
        iter.seek(SeekKey::Start);

        if iter.kv().is_none() {
            // FIXME return an error
        }

        let iter = iter.map(|(key, value)| {
            // retrieve the schema attribute name
            // from the schema attribute number
            let document_key_attr = DocumentKeyAttr::from_bytes(&key);
            let schema_attr = document_key_attr.attribute();
            let attribute_name = self.schema.attribute_name(schema_attr);
            (attribute_name, Value(value))
        });

        let map_deserializer = MapDeserializer::new(iter);
        visitor.visit_map(map_deserializer)
    }
}

struct Value(Vec<u8>);

impl<'de> IntoDeserializer<'de, DeserializerError> for Value {
    type Deserializer = Self;

    fn into_deserializer(self) -> Self::Deserializer {
        self
    }
}

macro_rules! forward_to_bincode_values {
    ($($ty:ident => $de_method:ident,)*) => {
        $(
            fn $de_method<V>(self, visitor: V) -> Result<V::Value, Self::Error>
                where V: de::Visitor<'de>
            {
                match bincode::deserialize::<$ty>(&self.0) {
                    Ok(val) => val.into_deserializer().$de_method(visitor),
                    Err(e) => Err(de::Error::custom(e)),
                }
            }
        )*
    }
}

impl<'de, 'a> de::Deserializer<'de> for Value {
    type Error = DeserializerError;

    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.0.into_deserializer().deserialize_any(visitor)
    }

    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.deserialize_string(visitor)
    }

    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        match bincode::deserialize::<String>(&self.0) {
            Ok(val) => val.into_deserializer().deserialize_string(visitor),
            Err(e) => Err(de::Error::custom(e)),
        }
    }

    fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        self.deserialize_byte_buf(visitor)
    }

    fn deserialize_byte_buf<V>(self, visitor: V) -> Result<V::Value, Self::Error>
    where V: Visitor<'de>
    {
        match bincode::deserialize::<Vec<u8>>(&self.0) {
            Ok(val) => val.into_deserializer().deserialize_byte_buf(visitor),
            Err(e) => Err(de::Error::custom(e)),
        }
    }

    forward_to_bincode_values! {
        char => deserialize_char,
        bool => deserialize_bool,

        u8  => deserialize_u8,
        u16 => deserialize_u16,
        u32 => deserialize_u32,
        u64 => deserialize_u64,

        i8  => deserialize_i8,
        i16 => deserialize_i16,
        i32 => deserialize_i32,
        i64 => deserialize_i64,

        f32 => deserialize_f32,
        f64 => deserialize_f64,
    }

    forward_to_deserialize_any! {
        unit seq map
        unit_struct tuple_struct
        identifier tuple ignored_any option newtype_struct enum struct
    }
}

#[derive(Debug)]
pub enum DeserializerError {
    Custom(String),
}

impl de::Error for DeserializerError {
    fn custom<T: fmt::Display>(msg: T) -> Self {
        DeserializerError::Custom(msg.to_string())
    }
}

impl fmt::Display for DeserializerError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            DeserializerError::Custom(s) => f.write_str(&s),
        }
    }
}

impl Error for DeserializerError {}
