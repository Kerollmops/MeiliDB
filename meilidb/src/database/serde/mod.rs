use std::collections::hash_map::DefaultHasher;
use std::hash::{Hash, Hasher};
use std::error::Error;
use std::fmt;

use serde::ser;

macro_rules! forward_to_unserializable_type {
    ($($ty:ident => $se_method:ident,)*) => {
        $(
            fn $se_method(self, _v: $ty) -> Result<Self::Ok, Self::Error> {
                Err(SerializerError::UnserializableType { name: "$ty" })
            }
        )*
    }
}

pub mod find_id;
pub mod key_to_string;
pub mod value_to_number;
pub mod serializer;
pub mod indexer_serializer;
pub mod deserializer;

pub fn calculate_hash<T: Hash>(t: &T) -> u64 {
    let mut s = DefaultHasher::new();
    t.hash(&mut s);
    s.finish()
}

#[derive(Debug)]
pub enum SerializerError {
    DocumentIdNotFound,
    UnserializableType { name: &'static str },
    Custom(String),
}

impl ser::Error for SerializerError {
    fn custom<T: fmt::Display>(msg: T) -> Self {
        SerializerError::Custom(msg.to_string())
    }
}

impl fmt::Display for SerializerError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            SerializerError::DocumentIdNotFound => {
                write!(f, "serialized document does not have an id according to the schema")
            }
            SerializerError::UnserializableType { name } => {
                write!(f, "Only struct and map types are considered valid documents and
                           can be serialized, not {} types directly.", name)
            },
            SerializerError::Custom(s) => f.write_str(&s),
        }
    }
}

impl Error for SerializerError {}

impl From<String> for SerializerError {
    fn from(value: String) -> SerializerError {
        SerializerError::Custom(value)
    }
}
