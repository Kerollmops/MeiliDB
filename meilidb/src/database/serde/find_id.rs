use serde::Serialize;
use serde::ser;

use crate::database::serde::key_to_string::KeyToStringSerializer;
use crate::database::serde::{SerializerError, calculate_hash};
use meilidb_core::DocumentId;

pub struct FindDocumentIdSerializer<'a> {
    pub id_attribute_name: &'a str,
}

impl<'a> ser::Serializer for FindDocumentIdSerializer<'a> {
    type Ok = DocumentId;
    type Error = SerializerError;
    type SerializeSeq = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeTuple = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeTupleStruct = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeTupleVariant = ser::Impossible<Self::Ok, Self::Error>;
    type SerializeMap = FindDocumentIdMapSerializer<'a>;
    type SerializeStruct = FindDocumentIdStructSerializer<'a>;
    type SerializeStructVariant = ser::Impossible<Self::Ok, Self::Error>;

    forward_to_unserializable_type! {
        bool => serialize_bool,
        char => serialize_char,

        i8  => serialize_i8,
        i16 => serialize_i16,
        i32 => serialize_i32,
        i64 => serialize_i64,

        u8  => serialize_u8,
        u16 => serialize_u16,
        u32 => serialize_u32,
        u64 => serialize_u64,

        f32 => serialize_f32,
        f64 => serialize_f64,
    }

    fn serialize_str(self, _v: &str) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "str" })
    }

    fn serialize_bytes(self, _v: &[u8]) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "&[u8]" })
    }

    fn serialize_none(self) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "Option" })
    }

    fn serialize_some<T: ?Sized>(self, _value: &T) -> Result<Self::Ok, Self::Error>
    where T: Serialize,
    {
        Err(SerializerError::UnserializableType { name: "Option" })
    }

    fn serialize_unit(self) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "()" })
    }

    fn serialize_unit_struct(self, _name: &'static str) -> Result<Self::Ok, Self::Error> {
        Err(SerializerError::UnserializableType { name: "unit struct" })
    }

    fn serialize_unit_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str
    ) -> Result<Self::Ok, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "unit variant" })
    }

    fn serialize_newtype_struct<T: ?Sized>(
        self,
        _name: &'static str,
        value: &T
    ) -> Result<Self::Ok, Self::Error>
    where T: Serialize,
    {
        value.serialize(self)
    }

    fn serialize_newtype_variant<T: ?Sized>(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str,
        _value: &T
    ) -> Result<Self::Ok, Self::Error>
    where T: Serialize,
    {
        Err(SerializerError::UnserializableType { name: "newtype variant" })
    }

    fn serialize_seq(self, _len: Option<usize>) -> Result<Self::SerializeSeq, Self::Error> {
        Err(SerializerError::UnserializableType { name: "sequence" })
    }

    fn serialize_tuple(self, _len: usize) -> Result<Self::SerializeTuple, Self::Error> {
        Err(SerializerError::UnserializableType { name: "tuple" })
    }

    fn serialize_tuple_struct(
        self,
        _name: &'static str,
        _len: usize
    ) -> Result<Self::SerializeTupleStruct, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "tuple struct" })
    }

    fn serialize_tuple_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str,
        _len: usize
    ) -> Result<Self::SerializeTupleVariant, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "tuple variant" })
    }

    fn serialize_map(self, _len: Option<usize>) -> Result<Self::SerializeMap, Self::Error> {
        Ok(FindDocumentIdMapSerializer {
            id_attribute_name: self.id_attribute_name,
            document_id: None,
            current_key_name: None,
        })
    }

    fn serialize_struct(
        self,
        _name: &'static str,
        _len: usize
    ) -> Result<Self::SerializeStruct, Self::Error>
    {
        Ok(FindDocumentIdStructSerializer {
            id_attribute_name: self.id_attribute_name,
            document_id: None,
        })
    }

    fn serialize_struct_variant(
        self,
        _name: &'static str,
        _variant_index: u32,
        _variant: &'static str,
        _len: usize
    ) -> Result<Self::SerializeStructVariant, Self::Error>
    {
        Err(SerializerError::UnserializableType { name: "struct variant" })
    }
}

pub struct FindDocumentIdMapSerializer<'a> {
    id_attribute_name: &'a str,
    document_id: Option<DocumentId>,
    current_key_name: Option<String>,
}

impl<'a> ser::SerializeMap for FindDocumentIdMapSerializer<'a> {
    type Ok = DocumentId;
    type Error = SerializerError;

    fn serialize_key<T: ?Sized>(&mut self, key: &T) -> Result<(), Self::Error>
    where T: Serialize,
    {
        let key = key.serialize(KeyToStringSerializer)?;
        self.current_key_name = Some(key);
        Ok(())
    }

    fn serialize_value<T: ?Sized>(&mut self, value: &T) -> Result<(), Self::Error>
    where T: Serialize,
    {
        let key = self.current_key_name.take().unwrap();
        self.serialize_entry(&key, value)
    }

    fn serialize_entry<K: ?Sized, V: ?Sized>(
        &mut self,
        key: &K,
        value: &V
    ) -> Result<(), Self::Error>
    where K: Serialize, V: Serialize,
    {
        let key = key.serialize(KeyToStringSerializer)?;

        if self.id_attribute_name == key {
            // TODO is it possible to have multiple ids?
            let id = bincode::serialize(value).unwrap();
            let hash = calculate_hash(&id);
            self.document_id = Some(DocumentId(hash));
        }

        Ok(())
    }

    fn end(self) -> Result<Self::Ok, Self::Error> {
        match self.document_id {
            Some(document_id) => Ok(document_id),
            None => Err(SerializerError::DocumentIdNotFound)
        }
    }
}

pub struct FindDocumentIdStructSerializer<'a> {
    id_attribute_name: &'a str,
    document_id: Option<DocumentId>,
}

impl<'a> ser::SerializeStruct for FindDocumentIdStructSerializer<'a> {
    type Ok = DocumentId;
    type Error = SerializerError;

    fn serialize_field<T: ?Sized>(
        &mut self,
        key: &'static str,
        value: &T
    ) -> Result<(), Self::Error>
    where T: Serialize,
    {
        if self.id_attribute_name == key {
            // TODO can it be possible to have multiple ids?
            let id = bincode::serialize(value).unwrap();
            let hash = calculate_hash(&id);
            self.document_id = Some(DocumentId(hash));
        }

        Ok(())
    }

    fn end(self) -> Result<Self::Ok, Self::Error> {
        match self.document_id {
            Some(document_id) => Ok(document_id),
            None => Err(SerializerError::DocumentIdNotFound)
        }
    }
}
