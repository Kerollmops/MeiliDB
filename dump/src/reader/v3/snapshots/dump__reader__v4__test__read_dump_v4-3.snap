---
source: dump/src/reader/v4/mod.rs
expression: tasks
---
[
  {
    "id": 9,
    "index_uid": "movies_2",
    "content": {
      "DocumentAddition": {
        "content_uuid": "3b12a971-bca2-4716-9889-36ffb715ae1d",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 200,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:49.125132233Z"
      }
    ]
  },
  {
    "id": 8,
    "index_uid": "movies",
    "content": {
      "DocumentAddition": {
        "content_uuid": "cae3205a-6016-471b-81de-081a195f098c",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 100,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:49.114226973Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:49.125918825Z",
          "batch_id": 8
        }
      },
      {
        "Processing": "2022-10-06T12:53:49.125930546Z"
      },
      {
        "Succeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 100
            }
          },
          "timestamp": "2022-10-06T12:53:49.785862546Z"
        }
      }
    ]
  },
  {
    "id": 7,
    "index_uid": "dnd_spells",
    "content": {
      "DocumentAddition": {
        "content_uuid": "7ba1eaa0-d2fb-4852-8d00-f35ed166728f",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "index",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:41.070732179Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:41.085563291Z",
          "batch_id": 7
        }
      },
      {
        "Processing": "2022-10-06T12:53:41.085563961Z"
      },
      {
        "Succeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-06T12:53:41.116036186Z"
        }
      }
    ]
  },
  {
    "id": 6,
    "index_uid": "dnd_spells",
    "content": {
      "DocumentAddition": {
        "content_uuid": "f2fb7d6e-11b6-45d9-aa7a-9495a567a275",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:40.831649057Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:40.834515892Z",
          "batch_id": 6
        }
      },
      {
        "Processing": "2022-10-06T12:53:40.834516572Z"
      },
      {
        "Failed": {
          "error": {
            "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
            "code": "primary_key_inference_failed",
            "type": "invalid_request",
            "link": "https://docs.meilisearch.com/errors#primary_key_inference_failed"
          },
          "timestamp": "2022-10-06T12:53:40.905384918Z"
        }
      }
    ]
  },
  {
    "id": 5,
    "index_uid": "products",
    "content": {
      "DocumentAddition": {
        "content_uuid": "f269fe46-36fe-4fe7-8c4e-2054f1b23594",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": "sku",
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:40.576727649Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:40.587595408Z",
          "batch_id": 5
        }
      },
      {
        "Processing": "2022-10-06T12:53:40.587596158Z"
      },
      {
        "Succeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-06T12:53:40.603035979Z"
        }
      }
    ]
  },
  {
    "id": 4,
    "index_uid": "products",
    "content": {
      "DocumentAddition": {
        "content_uuid": "7d1ea292-cdb6-4f47-8b25-c2ddde89035c",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:39.979427178Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:39.986159313Z",
          "batch_id": 4
        }
      },
      {
        "Processing": "2022-10-06T12:53:39.986160113Z"
      },
      {
        "Failed": {
          "error": {
            "message": "The primary key inference process failed because the engine did not find any fields containing `id` substring in their name. If your document identifier does not contain any `id` substring, you can set the primary key of the index.",
            "code": "primary_key_inference_failed",
            "type": "invalid_request",
            "link": "https://docs.meilisearch.com/errors#primary_key_inference_failed"
          },
          "timestamp": "2022-10-06T12:53:39.98921592Z"
        }
      }
    ]
  },
  {
    "id": 3,
    "index_uid": "products",
    "content": {
      "SettingsUpdate": {
        "settings": {
          "synonyms": {
            "android": [
              "phone",
              "smartphone"
            ],
            "iphone": [
              "phone",
              "smartphone"
            ],
            "phone": [
              "smartphone",
              "iphone",
              "android"
            ]
          }
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:39.360187055Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:39.371250258Z",
          "batch_id": 3
        }
      },
      {
        "Processing": "2022-10-06T12:53:39.371250918Z"
      },
      {
        "Processing": "2022-10-06T12:53:39.373988491Z"
      },
      {
        "Succeded": {
          "result": "Other",
          "timestamp": "2022-10-06T12:53:39.449840865Z"
        }
      }
    ]
  },
  {
    "id": 2,
    "index_uid": "movies",
    "content": {
      "SettingsUpdate": {
        "settings": {
          "rankingRules": [
            "words",
            "typo",
            "proximity",
            "attribute",
            "sort",
            "exactness",
            "release_date:asc"
          ]
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:39.143829637Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:39.154803808Z",
          "batch_id": 2
        }
      },
      {
        "Processing": "2022-10-06T12:53:39.154804558Z"
      },
      {
        "Processing": "2022-10-06T12:53:39.157501241Z"
      },
      {
        "Succeded": {
          "result": "Other",
          "timestamp": "2022-10-06T12:53:39.160263154Z"
        }
      }
    ]
  },
  {
    "id": 1,
    "index_uid": "movies",
    "content": {
      "SettingsUpdate": {
        "settings": {
          "filterableAttributes": [
            "genres",
            "id"
          ],
          "sortableAttributes": [
            "release_date"
          ]
        },
        "is_deletion": false,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:38.922837679Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:38.937712641Z",
          "batch_id": 1
        }
      },
      {
        "Processing": "2022-10-06T12:53:38.937713141Z"
      },
      {
        "Processing": "2022-10-06T12:53:38.940482335Z"
      },
      {
        "Succeded": {
          "result": "Other",
          "timestamp": "2022-10-06T12:53:38.953566059Z"
        }
      }
    ]
  },
  {
    "id": 0,
    "index_uid": "movies",
    "content": {
      "DocumentAddition": {
        "content_uuid": "cee1eef7-fadd-4970-93dc-25518655175f",
        "merge_strategy": "ReplaceDocuments",
        "primary_key": null,
        "documents_count": 10,
        "allow_index_creation": true
      }
    },
    "events": [
      {
        "Created": "2022-10-06T12:53:38.710611568Z"
      },
      {
        "Batched": {
          "timestamp": "2022-10-06T12:53:38.717455314Z",
          "batch_id": 0
        }
      },
      {
        "Processing": "2022-10-06T12:53:38.717456194Z"
      },
      {
        "Succeded": {
          "result": {
            "DocumentAddition": {
              "indexed_documents": 10
            }
          },
          "timestamp": "2022-10-06T12:53:38.811687295Z"
        }
      }
    ]
  }
]
